File: apps/etf-service/src/services/market/service.ts
import { HTTPException } from "hono/http-exception";
import type { AdjType, Security } from "../../common/schema";
import type { SecurityConfig } from "../../config/securities";
import type {
  DailyBar,
  GetDailyBarsRequest,
  GetDailyBarsResponse,
} from "../../routes/market/schema";
import { addDays, compareDate, isWeekend, listDates, maxDate, minDate } from "../../utils/date";

export interface MarketSecurityRecord {
  symbol: string;
  name: string;
  assetType: Security["assetType"];
  exchange: string | null;
  currency: string;
  source: string | null;
  earliestTradeDate: string;
}

export interface MarketBarInput {
  symbol: string;
  adjType?: AdjType;
  tradeDate: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number | null;
  amount?: number | null;
  changeAmount?: number | null;
  changePercent?: number | null;
  rawWeekday?: string | null;
}

export interface MarketCalendarInput {
  exchange: string;
  tradeDate: string;
  isOpen: number;
}

export interface MarketDataStore {
  upsertSecurity: (params: {
    security: SecurityConfig;
    earliestTradeDate: string;
  }) => Promise<void>;
  upsertDailyBars: (params: { bars: MarketBarInput[]; exchange: string }) => Promise<void>;
  upsertCalendarRows: (rows: MarketCalendarInput[]) => Promise<void>;
  listSecuritiesRows: () => Promise<MarketSecurityRecord[]>;
  getSecurityRow: (symbol: string) => Promise<MarketSecurityRecord | null>;
  getLatestCachedTradeDate: (params: {
    symbol: string;
    adjType: AdjType;
  }) => Promise<string | null>;
  listBars: (params: {
    symbol: string;
    adjType: AdjType;
    startDate: string;
    endDate: string;
  }) => Promise<DailyBar[]>;
  getCalendarMap: (params: {
    exchange: string;
    startDate: string;
    endDate: string;
  }) => Promise<Map<string, number>>;
}

export interface MarketService {
  upsertSecurity: (params: {
    security: SecurityConfig;
    earliestTradeDate: string;
  }) => Promise<void>;
  upsertDailyBars: (params: { bars: MarketBarInput[]; exchange: string }) => Promise<void>;
  listSecurities: () => Promise<Security[]>;
  getDailyBars: (params: GetDailyBarsRequest) => Promise<GetDailyBarsResponse>;
}

export interface CreateMarketServiceParams {
  securities: SecurityConfig[];
  store: MarketDataStore;
  fetchRemoteKlineBars: (params: { security: SecurityConfig }) => Promise<{
    bars: MarketBarInput[];
  }>;
  getTMinusOneDate: () => string;
}

const findSecurityConfig = (params: {
  securities: SecurityConfig[];
  symbol: string;
}): SecurityConfig => {
  const security = params.securities.find((item) => item.symbol === params.symbol);
  if (!security) {
    throw new HTTPException(404, { message: `Unknown security: ${params.symbol}` });
  }
  return security;
};

const toSecurity = (params: {
  row: MarketSecurityRecord;
  latestCachedTradeDate: string | null;
}): Security => ({
  symbol: params.row.symbol,
  name: params.row.name,
  assetType: params.row.assetType,
  exchange: params.row.exchange,
  currency: params.row.currency,
  source: params.row.source,
  earliestTradeDate: params.row.earliestTradeDate,
  latestCachedTradeDate: params.latestCachedTradeDate,
});

const getLatestRequiredOpenDate = async (params: {
  store: MarketDataStore;
  exchange: string;
  startDate: string;
  endDate: string;
}): Promise<string | null> => {
  const calendar = await params.store.getCalendarMap(params);
  const dates = listDates({ startDate: params.startDate, endDate: params.endDate }).reverse();

  for (const date of dates) {
    const calendarValue = calendar.get(date);
    if (calendarValue === 0 || isWeekend(date)) {
      continue;
    }
    return date;
  }

  return null;
};

const markClosedMissingDates = async (params: {
  store: MarketDataStore;
  exchange: string;
  startDate: string;
  endDate: string;
  bars: DailyBar[];
}): Promise<void> => {
  const openDates = new Set(params.bars.map((bar) => bar.tradeDate));
  const rows = listDates({ startDate: params.startDate, endDate: params.endDate })
    .filter((date) => !isWeekend(date) && !openDates.has(date))
    .map((tradeDate) => ({ exchange: params.exchange, tradeDate, isOpen: 0 }));

  await params.store.upsertCalendarRows(rows);
};

export const createMarketService = (params: CreateMarketServiceParams): MarketService => {
  const { securities, store, fetchRemoteKlineBars, getTMinusOneDate } = params;

  const listSecurities = async (): Promise<Security[]> => {
    const rows = await store.listSecuritiesRows();
    const result: Security[] = [];

    for (const row of rows) {
      const latestCachedTradeDate = await store.getLatestCachedTradeDate({
        symbol: row.symbol,
        adjType: "qfq",
      });
      result.push(toSecurity({ row, latestCachedTradeDate }));
    }

    return result;
  };

  const getDailyBars = async (request: GetDailyBarsRequest): Promise<GetDailyBarsResponse> => {
    const securityConfig = findSecurityConfig({ securities, symbol: request.symbol });
    const securityRow = await store.getSecurityRow(request.symbol);
    if (!securityRow) {
      throw new HTTPException(404, { message: `Unknown security: ${request.symbol}` });
    }

    const tMinusOne = getTMinusOneDate();
    const requestedStartDate = request.startDate ?? securityRow.earliestTradeDate;
    const requestedEndDate = request.endDate ?? tMinusOne;
    const effectiveStartDate = maxDate(requestedStartDate, securityRow.earliestTradeDate);
    const effectiveEndDate = minDate(requestedEndDate, tMinusOne);
    const latestCachedBefore = await store.getLatestCachedTradeDate(request);
    const security = toSecurity({ row: securityRow, latestCachedTradeDate: latestCachedBefore });

    if (compareDate(effectiveEndDate, effectiveStartDate) < 0) {
      return {
        security,
        bars: [],
        meta: {
          cacheStatus: "invalid",
          requestedStartDate,
          requestedEndDate,
          effectiveStartDate: null,
          effectiveEndDate: null,
          earliestTradeDate: securityRow.earliestTradeDate,
          latestCachedTradeDate: latestCachedBefore,
          refreshed: false,
          rows: 0,
        },
      };
    }

    const requiredOpenDate = await getLatestRequiredOpenDate({
      store,
      exchange: securityConfig.exchange,
      startDate: effectiveStartDate,
      endDate: effectiveEndDate,
    });
    const shouldRefresh =
      requiredOpenDate !== null &&
      (latestCachedBefore === null || compareDate(latestCachedBefore, requiredOpenDate) < 0);

    if (shouldRefresh) {
      const parsed = await fetchRemoteKlineBars({ security: securityConfig });
      const eligibleBars = parsed.bars.filter((bar) => compareDate(bar.tradeDate, tMinusOne) <= 0);
      await store.upsertDailyBars({ bars: eligibleBars, exchange: securityConfig.exchange });
    }

    const bars = await store.listBars({
      symbol: request.symbol,
      adjType: request.adjType,
      startDate: effectiveStartDate,
      endDate: effectiveEndDate,
    });

    if (shouldRefresh) {
      await markClosedMissingDates({
        store,
        exchange: securityConfig.exchange,
        startDate: addDays({ dateText: latestCachedBefore ?? effectiveStartDate, days: 1 }),
        endDate: effectiveEndDate,
        bars,
      });
    }

    const latestCachedTradeDate = await store.getLatestCachedTradeDate(request);

    return {
      security: toSecurity({ row: securityRow, latestCachedTradeDate }),
      bars,
      meta: {
        cacheStatus: shouldRefresh ? "refreshed" : "cache",
        requestedStartDate,
        requestedEndDate,
        effectiveStartDate,
        effectiveEndDate,
        earliestTradeDate: securityRow.earliestTradeDate,
        latestCachedTradeDate,
        refreshed: shouldRefresh,
        rows: bars.length,
      },
    };
  };

  return {
    upsertSecurity: store.upsertSecurity,
    upsertDailyBars: store.upsertDailyBars,
    listSecurities,
    getDailyBars,
  };
};

File: apps/etf-service/internal/modules/market/service.go
package market

import (
	"context"
	"fmt"
	"time"

	"kxh-awesome/etf-service/internal/shared/config"
	"kxh-awesome/etf-service/internal/shared/utils"
)

type MarketStore interface {
	UpsertDailyBars(ctx context.Context, bars []MarketBarInput, exchange string) error
	UpsertCalendarRows(ctx context.Context, rows []CalendarInput) error
	ListSecuritiesRows(ctx context.Context) ([]SecurityRecord, error)
	GetSecurityRow(ctx context.Context, symbol string) (*SecurityRecord, error)
	GetLatestCachedTradeDate(ctx context.Context, symbol string, adjType string) (*string, error)
	ListBars(ctx context.Context, symbol string, adjType string, startDate string, endDate string) ([]DailyBar, error)
	GetCalendarMap(ctx context.Context, exchange string, startDate string, endDate string) (map[string]int, error)
}

type RemoteFetcher interface {
	FetchRemoteKlineBars(ctx context.Context, security config.SecurityConfig) ([]MarketBarInput, error)
}

type MarketService struct {
	securities []config.SecurityConfig
	store      MarketStore
	fetcher    RemoteFetcher
	now        func() time.Time
}

func NewMarketService(securities []config.SecurityConfig, store MarketStore, fetcher RemoteFetcher) *MarketService {
	return &MarketService{
		securities: securities,
		store:      store,
		fetcher:    fetcher,
		now:        time.Now,
	}
}

func NewMarketServiceWithNow(securities []config.SecurityConfig, store MarketStore, fetcher RemoteFetcher, now func() time.Time) *MarketService {
	service := NewMarketService(securities, store, fetcher)
	service.now = now
	return service
}

func (s *MarketService) ListSecurities(ctx context.Context) ([]Security, error) {
	rows, err := s.store.ListSecuritiesRows(ctx)
	if err != nil {
		return nil, err
	}

	result := make([]Security, 0, len(rows))
	for _, row := range rows {
		latestCachedTradeDate, err := s.store.GetLatestCachedTradeDate(ctx, row.Symbol, "qfq")
		if err != nil {
			return nil, err
		}
		result = append(result, toSecurity(row, latestCachedTradeDate))
	}
	return result, nil
}

func (s *MarketService) GetDailyBars(ctx context.Context, request GetDailyBarsRequest) (*GetDailyBarsResponse, error) {
	if request.AdjType == "" {
		request.AdjType = "qfq"
	}

	securityConfig, ok := s.findSecurityConfig(request.Symbol)
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownSecurity, request.Symbol)
	}

	securityRow, err := s.store.GetSecurityRow(ctx, request.Symbol)
	if err != nil {
		return nil, err
	}
	if securityRow == nil {
		return nil, fmt.Errorf("%w: %s", ErrUnknownSecurity, request.Symbol)
	}

	// 行情源在交易日盘中可能返回未定稿数据，本地缓存只认 T-1 以保持图表口径稳定。
	tMinusOne, err := utils.TMinusOne(s.now())
	if err != nil {
		return nil, err
	}

	requestedStartDate := securityRow.EarliestTradeDate
	if request.StartDate != nil {
		requestedStartDate = *request.StartDate
	}
	requestedEndDate := tMinusOne
	if request.EndDate != nil {
		requestedEndDate = *request.EndDate
	}

	// 请求范围裁剪到已知历史和已完成交易日，避免返回无意义空洞或半日数据。
	effectiveStartDate := utils.Max(requestedStartDate, securityRow.EarliestTradeDate)
	effectiveEndDate := utils.Min(requestedEndDate, tMinusOne)
	latestCachedBefore, err := s.store.GetLatestCachedTradeDate(ctx, request.Symbol, request.AdjType)
	if err != nil {
		return nil, err
	}
	security := toSecurity(*securityRow, latestCachedBefore)

	if utils.Compare(effectiveEndDate, effectiveStartDate) < 0 {
		return &GetDailyBarsResponse{
			Security: security,
			Bars:     []DailyBar{},
			Meta: GetDailyBarsMeta{
				CacheStatus:           "invalid",
				RequestedStartDate:    &requestedStartDate,
				RequestedEndDate:      &requestedEndDate,
				EffectiveStartDate:    nil,
				EffectiveEndDate:      nil,
				EarliestTradeDate:     securityRow.EarliestTradeDate,
				LatestCachedTradeDate: latestCachedBefore,
				Refreshed:             false,
				Rows:                  0,
			},
		}, nil
	}

	// 刷新判断基于最近应开市日，防止周末和节假日把缓存误判为落后。
	requiredOpenDate, err := s.getLatestRequiredOpenDate(ctx, securityConfig.Exchange, effectiveStartDate, effectiveEndDate)
	if err != nil {
		return nil, err
	}

	shouldRefresh := requiredOpenDate != nil &&
		(latestCachedBefore == nil || utils.Compare(*latestCachedBefore, *requiredOpenDate) < 0)

	if shouldRefresh {
		remoteBars, err := s.fetcher.FetchRemoteKlineBars(ctx, securityConfig)
		if err != nil {
			return nil, err
		}

		eligibleBars := make([]MarketBarInput, 0, len(remoteBars))
		for _, bar := range remoteBars {
			// 远端可能带出当天或未来占位数据，入库前再次拦截以保护缓存口径。
			if utils.Compare(bar.TradeDate, tMinusOne) <= 0 {
				eligibleBars = append(eligibleBars, bar)
			}
		}
		if err := s.store.UpsertDailyBars(ctx, eligibleBars, securityConfig.Exchange); err != nil {
			return nil, err
		}
	}

	bars, err := s.store.ListBars(ctx, request.Symbol, request.AdjType, effectiveStartDate, effectiveEndDate)
	if err != nil {
		return nil, err
	}

	if shouldRefresh {
		startDate := effectiveStartDate
		if latestCachedBefore != nil {
			startDate, err = utils.AddDays(*latestCachedBefore, 1)
			if err != nil {
				return nil, err
			}
		}
		// 刷新后把无 K 线的工作日记为休市，后续请求才能跳过已验证的缺口。
		if err := s.markClosedMissingDates(ctx, securityConfig.Exchange, startDate, effectiveEndDate, bars); err != nil {
			return nil, err
		}
	}

	latestCachedTradeDate, err := s.store.GetLatestCachedTradeDate(ctx, request.Symbol, request.AdjType)
	if err != nil {
		return nil, err
	}

	return &GetDailyBarsResponse{
		Security: toSecurity(*securityRow, latestCachedTradeDate),
		Bars:     bars,
		Meta: GetDailyBarsMeta{
			CacheStatus:           cacheStatus(shouldRefresh),
			RequestedStartDate:    &requestedStartDate,
			RequestedEndDate:      &requestedEndDate,
			EffectiveStartDate:    &effectiveStartDate,
			EffectiveEndDate:      &effectiveEndDate,
			EarliestTradeDate:     securityRow.EarliestTradeDate,
			LatestCachedTradeDate: latestCachedTradeDate,
			Refreshed:             shouldRefresh,
			Rows:                  int32(len(bars)),
		},
	}, nil
}

func (s *MarketService) findSecurityConfig(symbol string) (config.SecurityConfig, bool) {
	for _, security := range s.securities {
		if security.Symbol == symbol {
			return security, true
		}
	}
	return config.SecurityConfig{}, false
}

func (s *MarketService) getLatestRequiredOpenDate(ctx context.Context, exchange string, startDate string, endDate string) (*string, error) {
	calendar, err := s.store.GetCalendarMap(ctx, exchange, startDate, endDate)
	if err != nil {
		return nil, err
	}

	dates, err := utils.ListDates(startDate, endDate)
	if err != nil {
		return nil, err
	}

	for index := len(dates) - 1; index >= 0; index-- {
		date := dates[index]
		if value, ok := calendar[date]; ok && value == 0 {
			continue
		}
		isWeekend, err := utils.IsWeekend(date)
		if err != nil {
			return nil, err
		}
		if isWeekend {
			continue
		}
		return &date, nil
	}
	return nil, nil
}

func (s *MarketService) markClosedMissingDates(ctx context.Context, exchange string, startDate string, endDate string, bars []DailyBar) error {
	openDates := map[string]struct{}{}
	for _, bar := range bars {
		openDates[bar.TradeDate] = struct{}{}
	}

	dates, err := utils.ListDates(startDate, endDate)
	if err != nil {
		return err
	}

	rows := make([]CalendarInput, 0, len(dates))
	for _, tradeDate := range dates {
		if _, ok := openDates[tradeDate]; ok {
			continue
		}
		isWeekend, err := utils.IsWeekend(tradeDate)
		if err != nil {
			return err
		}
		if isWeekend {
			continue
		}
		rows = append(rows, CalendarInput{
			Exchange:  exchange,
			TradeDate: tradeDate,
			IsOpen:    0,
		})
	}

	return s.store.UpsertCalendarRows(ctx, rows)
}

func toSecurity(row SecurityRecord, latestCachedTradeDate *string) Security {
	return Security{
		Symbol:                row.Symbol,
		Name:                  row.Name,
		AssetType:             row.AssetType,
		Exchange:              row.Exchange,
		Currency:              row.Currency,
		Source:                row.Source,
		EarliestTradeDate:     row.EarliestTradeDate,
		LatestCachedTradeDate: latestCachedTradeDate,
	}
}

func cacheStatus(refreshed bool) string {
	if refreshed {
		return "refreshed"
	}
	return "cache"
}
